File: Power Query/300-399/329/PQ 329.R
library(tidyverse)
library(readxl)

path = "Power Query/300-399/329/PQ_Challenge_329.xlsx"
input = read_excel(path, range = "A1:C12")
test  = read_excel(path, range = "E1:F18")

result = input %>%
  fill(everything()) %>%
  mutate(Capital = str_remove(Cities, " \\(C\\)"),
         IsCapital = str_detect(Cities, "\\(C\\)")) %>%
  summarise(
    Capital = ifelse(any(IsCapital), first(Capital[IsCapital]), "None"),
    `Other Cities` = ifelse(any(!IsCapital), paste(Cities[!IsCapital], collapse = ", "), "None"),
    .by = c(Country, State)
  ) %>%
  mutate(rn = row_number()) %>%
  pivot_longer(-rn, names_to = "Type", values_to = "City") %>%
  filter(!(Type == "Country" & duplicated(City))) %>%
  select(-rn) %>%
  replace_na(list(City = "None"))

colnames(result) = colnames(test)

all.equal(result, test)
# TRUE

File: Power Query/300-399/329/PQ 329.py
import pandas as pd

path = "300-399/329/PQ_Challenge_329.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=12).ffill()
test = pd.read_excel(path, usecols="E:F", nrows=18).fillna("None")

input["cap"] = input["Cities"].str.contains(r"\(C\)")
input["city"] = input["Cities"].str.replace(r" \(C\)", "", regex=True)

agg = (input.groupby(["Country", "State"], sort=False)
    .apply(lambda g: pd.Series({
        "Capital": g.loc[g.cap, "city"].drop_duplicates().iloc[0] if g.cap.any() else "None",
        "Other Cities": ", ".join(g.loc[~g.cap, "city"].drop_duplicates()) or "None"
    })).reset_index())

rows = []
for c in input["Country"].drop_duplicates():
    rows.append(pd.DataFrame({"Type": ["Country"], "City": [c]}))
    for s in input.loc[input["Country"].eq(c), "State"].drop_duplicates():
     r = agg.query("Country == @c and State == @s").iloc[0]
     rows.append(pd.DataFrame({
         "Type": ["State", "Capital", "Other Cities"],
         "City": [s, r["Capital"], r["Other Cities"]]
     }))

result = pd.concat(rows, ignore_index=True)
result.columns = test.columns
print(result.equals(test))
